RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,974 @@
+    // If both TSOSs agree, use only the one at vertex, as it uses more information. If they do not agree, search for seeds based on both.
+    double L2muonEta = l2->eta();
+    double absL2muonEta = std::abs(L2muonEta);
+
+    // make non-const copies of parameters
+    // (we want to override them if DNN evaluation is enabled)
+    unsigned int maxHitSeeds__ = maxHitSeeds_;
+    unsigned int maxHitDoubletSeeds__ = maxHitDoubletSeeds_;
+    unsigned int maxHitlessSeedsIP__ = maxHitlessSeedsIP_;
+    unsigned int maxHitlessSeedsMuS__ = maxHitlessSeedsMuS_;
+
+    float errorSFHitless = fixedErrorRescalingForHitless_;
+
+    // update strategy parameters by evaluating DNN
+    if (getStrategyFromDNN_) {
+      int nHBd(0), nHLIP(0), nHLMuS(0);
+      float SF = 1.0;
+
+      bool dnnSuccess_ = false;
+
+      // Put variables needed for DNN into an std::map
+      std::map<std::string, float> feature_map_ = getFeatureMap(l2, tsosAtIP, outerTkStateOutside);
+
+      if (useRegressor_) {
+        // Use regressor
+        std::tie(nHBd, nHLIP, nHLMuS, SF, dnnSuccess_) = evaluateRegressor(feature_map_,
+                                                                           tf_session_HB_,
+                                                                           metadata.get_child("HB"),
+                                                                           tf_session_HLIP_,
+                                                                           metadata.get_child("HLIP"),
+                                                                           tf_session_HLMuS_,
+                                                                           metadata.get_child("HLMuS"));
+      } else {
+        // Use classifier
+        if (std::abs(l2->eta()) < etaSplitForDnn_) {
+          // barrel
+          std::tie(nHBd, nHLIP, nHLMuS, SF, dnnSuccess_) =
+              evaluateClassifier(feature_map_, tf_session_barrel_, metadata.get_child("barrel"));
+        } else {
+          // endcap
+          std::tie(nHBd, nHLIP, nHLMuS, SF, dnnSuccess_) =
+              evaluateClassifier(feature_map_, tf_session_endcap_, metadata.get_child("endcap"));
+        }
+      }
+      if (!dnnSuccess_)
+        break;
+
+      maxHitSeeds__ = 0;
+      maxHitDoubletSeeds__ = nHBd;
+      maxHitlessSeedsIP__ = nHLIP;
+      maxHitlessSeedsMuS__ = nHLMuS;
+      errorSFHitless = SF;
+    }
+
+    numSeedsMade = 0;
+    hitlessSeedsMadeIP = 0;
+    hitlessSeedsMadeMuS = 0;
+    hitSeedsMade = 0;
+    hitDoubletSeedsMade = 0;
+
+    // BARREL
+    if (absL2muonEta < maxEtaForTOB_) {
+      layerCount = 0;
+      for (auto it = tob.rbegin(); it != tob.rend(); ++it) {
+        LogTrace("TSGForOIDNN") << "TSGForOIDNN::produce: looping in TOB layer " << layerCount << std::endl;
+
+        if (hitlessSeedsMadeIP < maxHitlessSeedsIP__ && numSeedsMade < maxSeeds_)
+          makeSeedsWithoutHits(**it,
```

Comment:
the code inside the loop is repeated 3 times for the barrel, and the two endcaps.
Please refactor to e.g. a lambda and reuse instead of copy-pasting about 40 lines of code in each case.

Reply:
Done!